app/services/ping.rb

```diff
@@ -63,7 +63,7 @@ def statuses_smart_proxies
         begin
           version = proxy.statuses[:version].version['version']
           features = proxy.statuses[:version].version['modules']
-          failed_features = proxy.statuses[:logs].logs.failed_modules
+          failed_features = (proxy.statuses[:logs].nil? ? {} : proxy.statuses[:logs].logs.failed_modules)
```

Comment:
Should this explicitly check for the feature?
```suggestion
          failed_features = (proxy.has_feature?('Logs') ? proxy.statuses[:logs].logs.failed_modules : {})
```

That may result in an additional database query.

It may be even better to explicitly check the failed modules since there's a direct API for this. No need to resort to the logs module; that was written before the V2 Features API was introduced. https://theforeman.org/2019/04/smart-proxy-capabilities-explained.html

Right now https://github.com/theforeman/foreman/blob/4e53859604bf0970b6f9b5c024069569979c4fae/app/models/smart_proxy.rb#L167-L180 explicitly rejects non-running features so it's discarding the information you actually want right there.

Reply:
So sort like:

    failed_features = ProxyAPI::V2::Features.new(:url => proxy.url).features.delete_if{|f,p| p['state'] == 'running'}

?

Reply:
Essentially yes, but I'd explicitly check for `failed`. To quote my own blog:
> The current known values for this are uninitialized, loaded, starting, running, disabled and failed

However, I will note I'm not too familiar with the ping part. Perhaps @tbrisker knows more.

Reply:
Hi @tbrisker , coud you please suggest here on a proper test of foreman-proxy status failure?

Reply:
Makes sense to me to check the the status of the features directly with the proxy api, and check for failed explicitly so e.g. disabled features aren't listed as failed.